// square/dist/esm/models/cashDrawerShift 2.js
import { array, lazy, nullable, object, optional, string, } from '../schema';
import { cashDrawerDeviceSchema } from './cashDrawerDevice';
import { moneySchema } from './money';
export const cashDrawerShiftSchema = object({
    id: ['id', optional(string())],
    state: ['state', optional(string())],
    openedAt: ['opened_at', optional(nullable(string()))],
    endedAt: ['ended_at', optional(nullable(string()))],
    closedAt: ['closed_at', optional(nullable(string()))],
    employeeIds: ['employee_ids', optional(nullable(array(string())))],
    openingEmployeeId: ['opening_employee_id', optional(nullable(string()))],
    endingEmployeeId: ['ending_employee_id', optional(nullable(string()))],
    closingEmployeeId: ['closing_employee_id', optional(nullable(string()))],
    description: ['description', optional(nullable(string()))],
    openedCashMoney: ['opened_cash_money', optional(lazy(() => moneySchema))],
    cashPaymentMoney: ['cash_payment_money', optional(lazy(() => moneySchema))],
    cashRefundsMoney: ['cash_refunds_money', optional(lazy(() => moneySchema))],
    cashPaidInMoney: ['cash_paid_in_money', optional(lazy(() => moneySchema))],
    cashPaidOutMoney: ['cash_paid_out_money', optional(lazy(() => moneySchema))],
    expectedCashMoney: ['expected_cash_money', optional(lazy(() => moneySchema))],
    closedCashMoney: ['closed_cash_money', optional(lazy(() => moneySchema))],
    device: ['device', optional(lazy(() => cashDrawerDeviceSchema))],
});
